### PR TITLE
(FM-7138) Add support for Access Lists

### PR DESCRIPTION
**Pull request by willmeek**

### CODE REVIEW

**Comment by shermdog** [comment]
I'm going to deep dive on this one hopefully Fri 7/27.  Off the top of my head I think we should be consistent with compound keys and use spaces instead of underscores.

**Comment by willmeek** [comment]
Thanks @tphoney @shermdog 

Changed the compound names to use spaces rather than underscores

File: lib/puppet/provider/ios_acl_entry/ios.rb [hunk 1]
```diff
@@ -0,0 +1,453 @@
...
+        instance[:dscp] = split_output.shift
+
+      elsif next_token == 'fin'
+        instance[:fin] = true
+      elsif next_token == 'fragments'
+        instance[:fragments] = true
+
+      elsif next_token == 'log'
+        instance[:log] = true
+      elsif next_token == 'log_input'
```
Comment (shermdog) [inline lib/puppet/provider/ios_acl_entry/ios.rb]: This token should be `log-input`

**Comment by shermdog** [comment]
Dynamic entries reek havok on the resource

```
Extended IP access list test
    4 Dynamic eek permit ip any 192.168.0.0 0.0.255.255 log
```
```Puppet
ios_acl_entry { 'test 4': 
  entry => 4,
  permission => 'Dynamic',
  access_list => 'test',
  ensure => 'present',
  protocol => 'eek',
  log => true,
}
```

**Comment by shermdog** [comment]
Reflexive entries also cause issues:
https://www.cisco.com/c/en/us/td/docs/ios/12_2/security/configuration/guide/fsecur_c/scfreflx.html
```
cat-2960g#sh run | begin ip access-list
ip access-list extended inboundfilters
 permit eigrp any any
 deny   icmp any any
 evaluate tcptraffic 
ip access-list extended outboundfilters
 permit tcp any any reflect tcptraffic
```
```
cat-2960g#show ip access-lists         
Extended IP access list inboundfilters
    10 permit eigrp any any
    20 deny icmp any any
    30 evaluate tcptraffic
Extended IP access list outboundfilters
    10 permit tcp any any reflect tcptraffic
Reflexive IP access list tcptraffic
```

```
[root@puppet ~]# puppet device -v -t cat-2960 --resource ios_access_list
Info: retrieving resource: ios_access_list from cat-2960 at file:///etc/puppetlabs/puppet/devices/cat-2960.yaml
Warning: Provider returned data that does not match the Type Schema for `ios_access_list[tcptraffic]`
 Value type mismatch:
    * access_list_type: Reflexive

ios_access_list { 'inboundfilters': 
  access_list_type => 'Extended',
  ensure => 'present',
}
ios_access_list { 'outboundfilters': 
  access_list_type => 'Extended',
  ensure => 'present',
}
ios_access_list { 'tcptraffic': 
  access_list_type => 'Reflexive',
  ensure => 'present',
}
```

File: lib/puppet/type/ios_acl_entry.rb [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+    entry:    {
+      type:    'Integer',
+      desc:    'Name. Used as sequence number <1-2147483647>',
+    },
+    permission:      {
+      type:    'Enum["permit", "deny"]',
+      desc:    'Specify packets to forward/reject',
+    },
+    protocol:    {
+      type:    'Optional[Enum["ahp","eigrp","esp","gre","icmp","igmp","ip","ipinip","nos","ospf","pcp","pim","tcp","udp"]]',
```
Comment (shermdog) [inline lib/puppet/type/ios_acl_entry.rb]: We may want to extend this to support an integer value as well.

```
Warning: Provider returned data that does not match the Type Schema for `ios_acl_entry[test 20]`
 Value type mismatch:
    * protocol: 8

cat-2960g#show ip access-lists         
Extended IP access list test
    10 permit udp any 0.0.0.2 255.255.255.0 log
    20 permit 8 host 1.1.1.1 any precedence routine tos max-reliability option traceroute log-input fragments
```

File: lib/puppet/provider/ios_acl_entry/ios.rb [hunk 3]
```diff
@@ -0,0 +1,453 @@
...
+  def set(context, changes)
+    changes.each do |name, change|
+      # What type of ACL are we using
+      acl_type = Puppet::Provider::IosAclEntry::IosAclEntry.type_of_access_list(context.device.run_command_enable_mode("show ip access-lists #{change[:should][:access_list]}"))
+
+      is = change.key?(:is) ? change[:is] : (get(context) || []).find { |key| key[:name] == name }
+      should = change[:should]
+      if should[:ensure].to_s == 'absent'
+        if Puppet::Provider::IosAclEntry::IosAclEntry.name_of_access_list(context.device.run_command_enable_mode("show ip access-lists #{change[:should][:access_list]}")).nil?
+          raise "ios_acl_entry delete requires parent ios_access_list #{change[:should][:access_list]} to be already present"
```
Comment (shermdog) [inline lib/puppet/provider/ios_acl_entry/ios.rb]: We may also want to look at using an autorequire for acl_entry -> access_list

**Comment by shermdog** [comment]
Looks like there are some issues with fairly basic lists as well

```
cat-2960g#sh run | begin ip access-list
ip access-list extended test
 permit udp 0.0.0.0 255.255.255.254 eq 1985 0.0.0.0 255.255.255.254 eq 1985
 deny   udp any eq 1985 any eq 1985
 deny   icmp any any fragments
 permit icmp 0.0.0.0 255.255.255.0 any
 deny   icmp any any
 deny   ip any any
 deny   ip any host 10.10.10.0
 deny   ip any host 10.10.10.255
 permit ip 0.0.0.0 255.255.255.0 any
 deny   ip any any log
```
```
cat-2960g#show ip access-lists
Extended IP access list test
    10 permit udp 0.0.0.0 255.255.255.254 eq 1985 0.0.0.0 255.255.255.254 eq 1985
    20 deny udp any eq 1985 any eq 1985
    30 deny icmp any any fragments
    40 permit icmp 0.0.0.0 255.255.255.0 any
    50 deny icmp any any
    60 deny ip any any
    70 deny ip any host 10.10.10.0
    80 deny ip any host 10.10.10.255
    90 permit ip 0.0.0.0 255.255.255.0 any
    10000 deny ip any any log
```

Some resources are returning funny
```
[root@puppet ~]# puppet device -v -t cat-2960 --resource ios_acl_entry
Info: retrieving resource: ios_acl_entry from cat-2960 at file:///etc/puppetlabs/puppet/devices/cat-2960.yaml
ios_acl_entry { 'test 10': 
  entry => 10,
  permission => 'permit',
  access_list => 'test',
  ensure => 'present',
  protocol => 'udp',
  source_address => '0.0.0.0',
  source_address_wildcard_mask => '255.255.255.254',
  source_eq => ['1985', '0.0.0.0', '255.255.255.254'],
}

ios_acl_entry { 'test 20': 
  entry => 20,
  permission => 'deny',
  access_list => 'test',
  ensure => 'present',
  protocol => 'udp',
  source_address_any => true,
  source_eq => ['1985'],
  destination_address_any => true,
  destination_eq => ['1985'],
}
```

**Comment by shermdog** [comment]
Also updating existing rules does not function.  I suspect we have to delete the existing rule then re-add it.
Existing rules
```Puppet
[root@puppet ~]# puppet device -v -t cat-2960 --resource ios_acl_entry
Info: retrieving resource: ios_acl_entry from cat-2960 at file:///etc/puppetlabs/puppet/devices/cat-2960.yaml
ios_acl_entry { 'test 20': 
  entry => 20,
  permission => 'deny',
  access_list => 'test',
  ensure => 'present',
  protocol => 'udp',
  source_address_any => true,
  source_eq => ['1985'],
  destination_address_any => true,
  destination_eq => ['1985'],
}
ios_acl_entry { 'test 30': 
  entry => 30,
  permission => 'permit',
  access_list => 'test',
  ensure => 'present',
  protocol => 'icmp',
  source_address => '0.0.0.0',
  source_address_wildcard_mask => '255.255.255.0',
  destination_address_any => true,
}
```

Update manifest
```Puppet
[root@puppet ~]# cat acl3.pp 
ios_access_list { 'test':
  access_list_type => 'Extended',
  ensure => 'present',
}


ios_acl_entry { 'test 30': 
  entry => 30,
  permission => 'deny',
  access_list => 'test',
  ensure => 'present',
  protocol => 'udp',
  source_address_any => true,
  source_eq => ['1985'],
  destination_address_any => true,
  destination_eq => ['1985'],
  require => Ios_access_list['test'],
}
ios_acl_entry { 'test 20': 
  entry => 20,
  permission => 'permit',
  access_list => 'test',
  ensure => 'present',
  protocol => 'icmp',
  source_address => '0.0.0.0',
  source_address_wildcard_mask => '255.255.255.0',
  destination_address_any => true,
  require => Ios_access_list['test'],
}
```

Update run (truncated)
```
[root@puppet ~]# puppet device -d -t cat-2960 --apply acl3.pp --trace
Notice: /Stage[main]/Main/Ios_acl_entry[test 30]/permission: permission changed 'permit' to 'deny'
Notice: /Stage[main]/Main/Ios_acl_entry[test 30]/protocol: protocol changed 'icmp' to 'udp'
Notice: /Stage[main]/Main/Ios_acl_entry[test 30]/source_address_any: source_address_any changed  to 'true'
Notice: /Stage[main]/Main/Ios_acl_entry[test 30]/source_eq: source_eq changed  to ['1985']
Notice: /Stage[main]/Main/Ios_acl_entry[test 30]/destination_eq: destination_eq changed  to ['1985']
Debug: ios_acl_entry does not support `canonicalize`
Debug: Target State: {:name=>"test 30", :ensure=>"present", :access_list=>"test", :entry=>30, :permission=>"deny", :protocol=>"udp", :source_address_any=>true, :source_eq=>["1985"], :destination_address_any=>true, :destination_eq=>["1985"]}
Debug: ios_acl_entry does not support `supports_noop`
Debug: cisco_ios.send_command from /etc/puppetlabs/code/environments/production/modules/cisco_ios/lib/puppet/provider/ios_acl_entry/ios.rb:418:in `block in set':'"show ip access-lists test\nExtended IP access list test\n    20 deny udp any eq 1985 any eq 1985\n    30 permit icmp 0.0.0.0 255.255.255.0 any\ncat-2960g#"'
Debug: cisco_ios.send_command from /etc/puppetlabs/code/environments/production/modules/cisco_ios/lib/puppet/provider/ios_acl_entry/ios.rb:416:in `each':'"show ip access-lists test\nExtended IP access list test\n    20 deny udp any eq 1985 any eq 1985\n    30 permit icmp 0.0.0.0 255.255.255.0 any\ncat-2960g#"'
Debug: cisco_ios.send_command from /etc/puppetlabs/code/environments/production/modules/cisco_ios/lib/puppet/provider/ios_acl_entry/ios.rb:416:in `set':'"show ip access-lists test\nExtended IP access list test\n    20 deny udp any eq 1985 any eq 1985\n    30 permit icmp 0.0.0.0 255.255.255.0 any\ncat-2960g#"'
Debug: ios_acl_entry[test 30]: Updating: Start
Debug: cisco_ios.send_command from /etc/puppetlabs/code/environments/production/modules/cisco_ios/lib/puppet/provider/ios_acl_entry/ios.rb:441:in `block in update':'"ip access-list Extended test\ncat-2960g(config-ext-nacl)#"'
Debug: cisco_ios.send_command from /etc/puppetlabs/code/environments/production/modules/cisco_ios/lib/puppet/provider/ios_acl_entry/ios.rb:440:in `each':'"ip access-list Extended test\ncat-2960g(config-ext-nacl)#"'
Debug: cisco_ios.send_command from /etc/puppetlabs/code/environments/production/modules/cisco_ios/lib/puppet/provider/ios_acl_entry/ios.rb:440:in `update':'"ip access-list Extended test\ncat-2960g(config-ext-nacl)#"'
Debug: cisco_ios.send_command from /etc/puppetlabs/code/environments/production/modules/cisco_ios/lib/puppet/provider/ios_acl_entry/ios.rb:431:in `block (2 levels) in set':'"ip access-list Extended test\ncat-2960g(config-ext-nacl)#"'
Debug: cisco_ios.send_command from /etc/puppetlabs/code/environments/production/modules/cisco_ios/lib/puppet/provider/ios_acl_entry/ios.rb:430:in `block in set':'"ip access-list Extended test\ncat-2960g(config-ext-nacl)#"'
Debug: cisco_ios.send_command from /etc/puppetlabs/code/environments/production/modules/cisco_ios/lib/puppet/provider/ios_acl_entry/ios.rb:416:in `each':'"ip access-list Extended test\ncat-2960g(config-ext-nacl)#"'
Debug: cisco_ios.send_command from /etc/puppetlabs/code/environments/production/modules/cisco_ios/lib/puppet/provider/ios_acl_entry/ios.rb:416:in `set':'"ip access-list Extended test\ncat-2960g(config-ext-nacl)#"'
Debug: cisco_ios.send_command from /etc/puppetlabs/code/environments/production/modules/cisco_ios/lib/puppet/provider/ios_acl_entry/ios.rb:441:in `block in update':'" \ncat-2960g(config-ext-nacl)#"'
Debug: cisco_ios.send_command from /etc/puppetlabs/code/environments/production/modules/cisco_ios/lib/puppet/provider/ios_acl_entry/ios.rb:440:in `each':'" \ncat-2960g(config-ext-nacl)#"'
Debug: cisco_ios.send_command from /etc/puppetlabs/code/environments/production/modules/cisco_ios/lib/puppet/provider/ios_acl_entry/ios.rb:440:in `update':'" \ncat-2960g(config-ext-nacl)#"'
Debug: cisco_ios.send_command from /etc/puppetlabs/code/environments/production/modules/cisco_ios/lib/puppet/provider/ios_acl_entry/ios.rb:431:in `block (2 levels) in set':'" \ncat-2960g(config-ext-nacl)#"'
Debug: cisco_ios.send_command from /etc/puppetlabs/code/environments/production/modules/cisco_ios/lib/puppet/provider/ios_acl_entry/ios.rb:430:in `block in set':'" \ncat-2960g(config-ext-nacl)#"'
Debug: cisco_ios.send_command from /etc/puppetlabs/code/environments/production/modules/cisco_ios/lib/puppet/provider/ios_acl_entry/ios.rb:416:in `each':'" \ncat-2960g(config-ext-nacl)#"'
Debug: cisco_ios.send_command from /etc/puppetlabs/code/environments/production/modules/cisco_ios/lib/puppet/provider/ios_acl_entry/ios.rb:416:in `set':'" \ncat-2960g(config-ext-nacl)#"'
Debug: cisco_ios.send_command from /etc/puppetlabs/code/environments/production/modules/cisco_ios/lib/puppet/provider/ios_acl_entry/ios.rb:441:in `block in update':'"30 deny udp any eq 1985 any eq 1985\n% Duplicate sequence number.\n\ncat-2960g(config-ext-nacl)#"'
Debug: cisco_ios.send_command from /etc/puppetlabs/code/environments/production/modules/cisco_ios/lib/puppet/provider/ios_acl_entry/ios.rb:440:in `each':'"30 deny udp any eq 1985 any eq 1985\n% Duplicate sequence number.\n\ncat-2960g(config-ext-nacl)#"'
Debug: cisco_ios.send_command from /etc/puppetlabs/code/environments/production/modules/cisco_ios/lib/puppet/provider/ios_acl_entry/ios.rb:440:in `update':'"30 deny udp any eq 1985 any eq 1985\n% Duplicate sequence number.\n\ncat-2960g(config-ext-nacl)#"'
Debug: cisco_ios.send_command from /etc/puppetlabs/code/environments/production/modules/cisco_ios/lib/puppet/provider/ios_acl_entry/ios.rb:431:in `block (2 levels) in set':'"30 deny udp any eq 1985 any eq 1985\n% Duplicate sequence number.\n\ncat-2960g(config-ext-nacl)#"'
Debug: cisco_ios.send_command from /etc/puppetlabs/code/environments/production/modules/cisco_ios/lib/puppet/provider/ios_acl_entry/ios.rb:430:in `block in set':'"30 deny udp any eq 1985 any eq 1985\n% Duplicate sequence number.\n\ncat-2960g(config-ext-nacl)#"'
Debug: cisco_ios.send_command from /etc/puppetlabs/code/environments/production/modules/cisco_ios/lib/puppet/provider/ios_acl_entry/ios.rb:416:in `each':'"30 deny udp any eq 1985 any eq 1985\n% Duplicate sequence number.\n\ncat-2960g(config-ext-nacl)#"'
Debug: cisco_ios.send_command from /etc/puppetlabs/code/environments/production/modules/cisco_ios/lib/puppet/provider/ios_acl_entry/ios.rb:416:in `set':'"30 deny udp any eq 1985 any eq 1985\n% Duplicate sequence number.\n\ncat-2960g(config-ext-nacl)#"'
Debug: cisco_ios.send_command from /etc/puppetlabs/code/environments/production/modules/cisco_ios/lib/puppet/provider/ios_acl_entry/ios.rb:441:in `block in update':'"exit\ncat-2960g(config)#"'
Debug: cisco_ios.send_command from /etc/puppetlabs/code/environments/production/modules/cisco_ios/lib/puppet/provider/ios_acl_entry/ios.rb:440:in `each':'"exit\ncat-2960g(config)#"'
Debug: cisco_ios.send_command from /etc/puppetlabs/code/environments/production/modules/cisco_ios/lib/puppet/provider/ios_acl_entry/ios.rb:440:in `update':'"exit\ncat-2960g(config)#"'
Debug: cisco_ios.send_command from /etc/puppetlabs/code/environments/production/modules/cisco_ios/lib/puppet/provider/ios_acl_entry/ios.rb:431:in `block (2 levels) in set':'"exit\ncat-2960g(config)#"'
Debug: cisco_ios.send_command from /etc/puppetlabs/code/environments/production/modules/cisco_ios/lib/puppet/provider/ios_acl_entry/ios.rb:430:in `block in set':'"exit\ncat-2960g(config)#"'
Debug: cisco_ios.send_command from /etc/puppetlabs/code/environments/production/modules/cisco_ios/lib/puppet/provider/ios_acl_entry/ios.rb:416:in `each':'"exit\ncat-2960g(config)#"'
Debug: cisco_ios.send_command from /etc/puppetlabs/code/environments/production/modules/cisco_ios/lib/puppet/provider/ios_acl_entry/ios.rb:416:in `set':'"exit\ncat-2960g(config)#"'


Notice: /Stage[main]/Main/Ios_acl_entry[test 20]/permission: permission changed 'deny' to 'permit'
Notice: /Stage[main]/Main/Ios_acl_entry[test 20]/protocol: protocol changed 'udp' to 'icmp'
Notice: /Stage[main]/Main/Ios_acl_entry[test 20]/source_address: source_address changed  to '0.0.0.0'
Notice: /Stage[main]/Main/Ios_acl_entry[test 20]/source_address_wildcard_mask: source_address_wildcard_mask changed  to '255.255.255.0'
Debug: ios_acl_entry does not support `canonicalize`
Debug: Target State: {:name=>"test 20", :ensure=>"present", :access_list=>"test", :entry=>20, :permission=>"permit", :protocol=>"icmp", :source_address=>"0.0.0.0", :source_address_wildcard_mask=>"255.255.255.0", :destination_address_any=>true}
Debug: ios_acl_entry does not support `supports_noop`
Debug: cisco_ios.send_command from /etc/puppetlabs/code/environments/production/modules/cisco_ios/lib/puppet/provider/ios_acl_entry/ios.rb:418:in `block in set':'"show ip access-lists test\nExtended IP access list test\n    20 deny udp any eq 1985 any eq 1985\n    30 permit icmp 0.0.0.0 255.255.255.0 any\ncat-2960g#"'
Debug: cisco_ios.send_command from /etc/puppetlabs/code/environments/production/modules/cisco_ios/lib/puppet/provider/ios_acl_entry/ios.rb:416:in `each':'"show ip access-lists test\nExtended IP access list test\n    20 deny udp any eq 1985 any eq 1985\n    30 permit icmp 0.0.0.0 255.255.255.0 any\ncat-2960g#"'
Debug: cisco_ios.send_command from /etc/puppetlabs/code/environments/production/modules/cisco_ios/lib/puppet/provider/ios_acl_entry/ios.rb:416:in `set':'"show ip access-lists test\nExtended IP access list test\n    20 deny udp any eq 1985 any eq 1985\n    30 permit icmp 0.0.0.0 255.255.255.0 any\ncat-2960g#"'
Debug: ios_acl_entry[test 20]: Updating: Start
Debug: cisco_ios.send_command from /etc/puppetlabs/code/environments/production/modules/cisco_ios/lib/puppet/provider/ios_acl_entry/ios.rb:441:in `block in update':'"ip access-list Extended test\ncat-2960g(config-ext-nacl)#"'
Debug: cisco_ios.send_command from /etc/puppetlabs/code/environments/production/modules/cisco_ios/lib/puppet/provider/ios_acl_entry/ios.rb:440:in `each':'"ip access-list Extended test\ncat-2960g(config-ext-nacl)#"'
Debug: cisco_ios.send_command from /etc/puppetlabs/code/environments/production/modules/cisco_ios/lib/puppet/provider/ios_acl_entry/ios.rb:440:in `update':'"ip access-list Extended test\ncat-2960g(config-ext-nacl)#"'
Debug: cisco_ios.send_command from /etc/puppetlabs/code/environments/production/modules/cisco_ios/lib/puppet/provider/ios_acl_entry/ios.rb:431:in `block (2 levels) in set':'"ip access-list Extended test\ncat-2960g(config-ext-nacl)#"'
Debug: cisco_ios.send_command from /etc/puppetlabs/code/environments/production/modules/cisco_ios/lib/puppet/provider/ios_acl_entry/ios.rb:430:in `block in set':'"ip access-list Extended test\ncat-2960g(config-ext-nacl)#"'
Debug: cisco_ios.send_command from /etc/puppetlabs/code/environments/production/modules/cisco_ios/lib/puppet/provider/ios_acl_entry/ios.rb:416:in `each':'"ip access-list Extended test\ncat-2960g(config-ext-nacl)#"'
Debug: cisco_ios.send_command from /etc/puppetlabs/code/environments/production/modules/cisco_ios/lib/puppet/provider/ios_acl_entry/ios.rb:416:in `set':'"ip access-list Extended test\ncat-2960g(config-ext-nacl)#"'
Debug: cisco_ios.send_command from /etc/puppetlabs/code/environments/production/modules/cisco_ios/lib/puppet/provider/ios_acl_entry/ios.rb:441:in `block in update':'" \ncat-2960g(config-ext-nacl)#"'
Debug: cisco_ios.send_command from /etc/puppetlabs/code/environments/production/modules/cisco_ios/lib/puppet/provider/ios_acl_entry/ios.rb:440:in `each':'" \ncat-2960g(config-ext-nacl)#"'
Debug: cisco_ios.send_command from /etc/puppetlabs/code/environments/production/modules/cisco_ios/lib/puppet/provider/ios_acl_entry/ios.rb:440:in `update':'" \ncat-2960g(config-ext-nacl)#"'
Debug: cisco_ios.send_command from /etc/puppetlabs/code/environments/production/modules/cisco_ios/lib/puppet/provider/ios_acl_entry/ios.rb:431:in `block (2 levels) in set':'" \ncat-2960g(config-ext-nacl)#"'
Debug: cisco_ios.send_command from /etc/puppetlabs/code/environments/production/modules/cisco_ios/lib/puppet/provider/ios_acl_entry/ios.rb:430:in `block in set':'" \ncat-2960g(config-ext-nacl)#"'
Debug: cisco_ios.send_command from /etc/puppetlabs/code/environments/production/modules/cisco_ios/lib/puppet/provider/ios_acl_entry/ios.rb:416:in `each':'" \ncat-2960g(config-ext-nacl)#"'
Debug: cisco_ios.send_command from /etc/puppetlabs/code/environments/production/modules/cisco_ios/lib/puppet/provider/ios_acl_entry/ios.rb:416:in `set':'" \ncat-2960g(config-ext-nacl)#"'
Debug: cisco_ios.send_command from /etc/puppetlabs/code/environments/production/modules/cisco_ios/lib/puppet/provider/ios_acl_entry/ios.rb:441:in `block in update':'"20 permit icmp 0.0.0.0 255.255.255.0 any\n% Duplicate sequence number.\n\ncat-2960g(config-ext-nacl)#"'
Debug: cisco_ios.send_command from /etc/puppetlabs/code/environments/production/modules/cisco_ios/lib/puppet/provider/ios_acl_entry/ios.rb:440:in `each':'"20 permit icmp 0.0.0.0 255.255.255.0 any\n% Duplicate sequence number.\n\ncat-2960g(config-ext-nacl)#"'
Debug: cisco_ios.send_command from /etc/puppetlabs/code/environments/production/modules/cisco_ios/lib/puppet/provider/ios_acl_entry/ios.rb:440:in `update':'"20 permit icmp 0.0.0.0 255.255.255.0 any\n% Duplicate sequence number.\n\ncat-2960g(config-ext-nacl)#"'
Debug: cisco_ios.send_command from /etc/puppetlabs/code/environments/production/modules/cisco_ios/lib/puppet/provider/ios_acl_entry/ios.rb:431:in `block (2 levels) in set':'"20 permit icmp 0.0.0.0 255.255.255.0 any\n% Duplicate sequence number.\n\ncat-2960g(config-ext-nacl)#"'
Debug: cisco_ios.send_command from /etc/puppetlabs/code/environments/production/modules/cisco_ios/lib/puppet/provider/ios_acl_entry/ios.rb:430:in `block in set':'"20 permit icmp 0.0.0.0 255.255.255.0 any\n% Duplicate sequence number.\n\ncat-2960g(config-ext-nacl)#"'
Debug: cisco_ios.send_command from /etc/puppetlabs/code/environments/production/modules/cisco_ios/lib/puppet/provider/ios_acl_entry/ios.rb:416:in `each':'"20 permit icmp 0.0.0.0 255.255.255.0 any\n% Duplicate sequence number.\n\ncat-2960g(config-ext-nacl)#"'
Debug: cisco_ios.send_command from /etc/puppetlabs/code/environments/production/modules/cisco_ios/lib/puppet/provider/ios_acl_entry/ios.rb:416:in `set':'"20 permit icmp 0.0.0.0 255.255.255.0 any\n% Duplicate sequence number.\n\ncat-2960g(config-ext-nacl)#"'
Debug: cisco_ios.send_command from /etc/puppetlabs/code/environments/production/modules/cisco_ios/lib/puppet/provider/ios_acl_entry/ios.rb:441:in `block in update':'"exit\ncat-2960g(config)#"'
Debug: cisco_ios.send_command from /etc/puppetlabs/code/environments/production/modules/cisco_ios/lib/puppet/provider/ios_acl_entry/ios.rb:440:in `each':'"exit\ncat-2960g(config)#"'
Debug: cisco_ios.send_command from /etc/puppetlabs/code/environments/production/modules/cisco_ios/lib/puppet/provider/ios_acl_entry/ios.rb:440:in `update':'"exit\ncat-2960g(config)#"'
Debug: cisco_ios.send_command from /etc/puppetlabs/code/environments/production/modules/cisco_ios/lib/puppet/provider/ios_acl_entry/ios.rb:431:in `block (2 levels) in set':'"exit\ncat-2960g(config)#"'
Debug: cisco_ios.send_command from /etc/puppetlabs/code/environments/production/modules/cisco_ios/lib/puppet/provider/ios_acl_entry/ios.rb:430:in `block in set':'"exit\ncat-2960g(config)#"'
Debug: cisco_ios.send_command from /etc/puppetlabs/code/environments/production/modules/cisco_ios/lib/puppet/provider/ios_acl_entry/ios.rb:416:in `each':'"exit\ncat-2960g(config)#"'
Debug: cisco_ios.send_command from /etc/puppetlabs/code/environments/production/modules/cisco_ios/lib/puppet/provider/ios_acl_entry/ios.rb:416:in `set':'"exit\ncat-2960g(config)#"'
Notice: ios_acl_entry[test 20]: Updating: Finished in 0.226932 seconds
```

Resultant resources (same as before)
```Puppet
[root@puppet ~]# puppet device -v -t cat-2960 --resource ios_acl_entry
Info: retrieving resource: ios_acl_entry from cat-2960 at file:///etc/puppetlabs/puppet/devices/cat-2960.yaml
ios_acl_entry { 'test 20': 
  entry => 20,
  permission => 'deny',
  access_list => 'test',
  ensure => 'present',
  protocol => 'udp',
  source_address_any => true,
  source_eq => ['1985'],
  destination_address_any => true,
  destination_eq => ['1985'],
}
ios_acl_entry { 'test 30': 
  entry => 30,
  permission => 'permit',
  access_list => 'test',
  ensure => 'present',
  protocol => 'icmp',
  source_address => '0.0.0.0',
  source_address_wildcard_mask => '255.255.255.0',
  destination_address_any => true,
}
```

File: lib/puppet/provider/ios_acl_entry/ios.rb (hunk 1 above)
Comment (willmeek) [reply lib/puppet/provider/ios_acl_entry/ios.rb]: Good catch. Amended a unit test to include this.

**Comment by willmeek** [comment]
I wasn't fully aware of Dynamic entries, having multiple entries per entry throws a bit of a spanner in the works in terms of modelling as a Puppet manifest...

**Comment by shermdog** [comment]
Dynamic entries in  ```show ip access-lists```  contain both the explicitly configured rule + dynamically generated state.  (Being a show command it really is all state information, but that's a detail)

The manifest should only model the configuration information.

```
Router#show ip access-lists 120
Extended IP access list 120
    10 Dynamic testlist permit ip any any log   <- Config info to be modeled
       permit ip host 171.68.109.64 any log  <- Ephemeral state - do not model
       permit ip host 171.68.109.158 any log <- Ephemeral state - do not model
    20 permit tcp any host 171.68.117.189 eq telnet (288 matches)
```

**Comment by willmeek** [comment]
@shermdog  I was concerned earlier that we could have multiple per dynamic list, which wouldn't be straightforward to apply.

Then did a bit of research and found, as you point out, that the dynamic entries are temporary state.

I was working on a solution which could've brought us those back with an appended form and a read-only dynamically generated field, with a name eg. `test120 10 dynamic 2`

However it is straightforward to ignore these if that is what we want. Thanks!

File: lib/puppet/type/ios_acl_entry.rb (hunk 2 above)
Comment (willmeek) [reply lib/puppet/type/ios_acl_entry.rb]: Added a variant regex in the type to allow an integer protocol

File: lib/puppet/provider/ios_acl_entry/ios.rb (hunk 3 above)
Comment (willmeek) [reply lib/puppet/provider/ios_acl_entry/ios.rb]: Can we discuss this - the acl_entry would also need to know the *type* of the ACL (eg. extended, standard, reflexive), is this something we want to add to the acl_entry type?

**Comment by willmeek** [comment]
The issues with the access lists

`cat-2960g#show ip access-lists
Extended IP access list test
    10 permit udp 0.0.0.0 255.255.255.254 eq 1985 0.0.0.0 255.255.255.254 eq 1985
    20 deny udp any eq 1985 any eq 1985
    30 deny icmp any any fragments
    40 permit icmp 0.0.0.0 255.255.255.0 any
    50 deny icmp any any
    60 deny ip any any
    70 deny ip any host 10.10.10.0
    80 deny ip any host 10.10.10.255
    90 permit ip 0.0.0.0 255.255.255.0 any
    10000 deny ip any any log`

Entry 10 - this was a defect due to the 'eq' regex being too lenient - it was matching on the next token being a digit, however this also picked up IP address octets, and confused matters. This has been tightened up.

Entry 20 - I don't see an issue?

**Comment by shermdog** [comment]
Line 20 was just for formatting :)

File: lib/puppet/provider/ios_acl_entry/ios.rb (hunk 3 above)
Comment (willmeek) [reply lib/puppet/provider/ios_acl_entry/ios.rb]: Done

**Comment by willmeek** [comment]
@shermdog I added logic to the update to remove and add.

However there is some behaviour that is evident if you run your tests 'as is' - you might notice that only 'test 20' comes back.

This is because the 'new' 'test 30' (first entry in the manifest) has the same entry content as the existing 'test 20' entry. The ACL logic seems to be smart enough to match that this access list entry already exists and doesn't add. It also does not show any message at all for us to catch :)

The module itself deals with individual atomic items in isolation, without foresight or hindsight (as in, being able to 'delete all' before execution)

2 workarounds
- Absent the second on the list on the manifest before it
- Run the manifest again. As the entry doesn't exist it will be created.

**Comment by willmeek** [comment]
@shermdog This should be ready for re review, thanks for kicking the tyres brilliantly :)

**Comment by willmeek** [comment]
port type issue - this was because the port types we parse only had the TCP ports. Added UDP ports.

**Comment by willmeek** [comment]
@shermdog Added UDP port types.

Added support for ICMP message type/code

Added support for IGMP message type.

**Comment by willmeek** [comment]
@shermdog I missed the special characters regex update, that should be in there now with unit tests.

UDP / ICMP / IGMP ports should've been in there too?

File: lib/puppet/provider/ios_access_list/command.yaml [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+get_values:
+  default: 'show ip access-lists'
+get_instances:
+  default: '\S* IP access list (?:(?:.| |\n )*\n)'
+set_values:
+  default: '<state>ip access-list <access_list_type> <name>'
+attributes:
+  name:
+    default:
+      get_value: '.*IP access list (\w+)'
```
[the remark this answers is not in the document]
Comment (willmeek) [reply lib/puppet/provider/ios_access_list/command.yaml]: Updated regex, added unit test to prove out.

File: lib/puppet/provider/ios_acl_entry/ios.rb [hunk 5]
```diff
@@ -0,0 +1,488 @@
...
+
+    instance
+  end
+
+  def self.type_of_access_list(output)
+    output.scan(%r{(\S*) IP access list.*}).flatten.first
+  end
+
+  def self.name_of_access_list(output)
+    output.scan(%r{.*IP access list (\w+)}).flatten.first
```
[the remark this answers is not in the document]
Comment (willmeek) [reply lib/puppet/provider/ios_acl_entry/ios.rb]: Updated regex, added unit test to prove out.

**Comment by shermdog** [comment]
I think this is good to go!  Great job @willmeek - this was a monster!

**Comment by willmeek** [comment]
Rebased from latest master

**Comment by Thomas-Franklin** [comment]
```
Warning: Provider returned data that does not match the Type Schema for `ios_acl_entry[test 30]`
 Value type mismatch:
    * icmp_message_code: 43
```

Seems to be returning strings but expecting integers, other than this I am happy to merge

**Comment by willmeek** [comment]
> ```
> Warning: Provider returned data that does not match the Type Schema for `ios_acl_entry[test 30]`
>  Value type mismatch:
>     * icmp_message_code: 43
> ```
> Seems to be returning strings but expecting integers, other than this I am happy to merge

Which in turn will get fixed by
https://github.com/puppetlabs/cisco_ios/pull/230
:)

**Pull request merged**